File: ds/2.信息结构/array_queue.c
// Copyright (c) 2022 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by General Public License that can be found
// in the LICENSE file.

#include "queue.h"

#include <assert.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

struct queue_s {
  element_type* data;
  ssize_t front;
  ssize_t rear;
  size_t capacity;
  size_t size;
};

queue_t* queue_new(size_t size) {
  queue_t* queue = malloc(sizeof(queue_t));
  assert(queue != NULL);
  queue->front = -1;
  queue->rear = -1;
  queue->capacity = size;
  queue->size = 0;
  const size_t data_size = sizeof(element_type) * size;
  queue->data = malloc(data_size);
  assert(queue->data != NULL);
  memset(queue->data, 0, data_size);

  return queue;
}

void queue_free(queue_t* queue) {
  free(queue->data);
  queue->data = NULL;
  free(queue);
}

bool queue_is_full(queue_t* queue) {
  return queue->size == queue->capacity;
}

bool queue_is_empty(queue_t* queue) {
  return queue->size == 0;
}

size_t queue_size(queue_t* queue) {
  return queue->size;
}

void queue_add(queue_t* queue, element_type value) {
  assert(!queue_is_full(queue));
  queue->rear = (queue->rear + 1) % (ssize_t)queue->capacity;
  queue->data[queue->rear] = value;
  queue->size += 1;
}

element_type queue_delete(queue_t* queue) {
  assert(!queue_is_empty(queue));
  queue->front = (queue->front + 1) % (ssize_t)queue->capacity;
  queue->size -= 1;
  return queue->data[queue->front];
}

element_type queue_get(queue_t* queue) {
  assert(!queue_is_empty(queue));
  return queue->data[queue->front+1];
}

void queue_debug_print(queue_t* queue) {
  if (queue_is_empty(queue)) {
    return;
  }

  if (queue->rear > queue->front) {
    for (ssize_t i = queue->front + 1; i <= queue->rear; ++i) {
      printf("%d, ", queue->data[i]);
    }
    printf("\n");
  } else {
    for (ssize_t i = queue->front + 1; i != queue->size; ++i) {
      printf("%d, ", queue->data[i]);
    }
    for (ssize_t i = 0; i <= queue->rear; ++i) {
      printf("%d, ", queue->data[i]);
    }
    printf("\n");
  }

}

File: queue/src/array_queue.rs
// Copyright (c) 2024 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by General Public License that can be found
// in the LICENSE file.

use std::cmp::Ordering;
use std::fmt;
use std::hash::{Hash, Hasher};

pub struct ArrayQueue<T> {
    len: usize,
    buf: Box<[Option<T>]>,
}

impl<T> ArrayQueue<T> {
    #[must_use]
    pub fn new(capacity: usize) -> Self {
        let values: Vec<Option<T>> = (0..capacity).map(|_| None).collect();

        Self {
            len: 0,
            buf: values.into_boxed_slice(),
        }
    }

    /// # Errors
    ///
    /// 当栈已满时再将元素入队, 就会返回错误, 以及原有的元素 `value`.
    pub fn push(&mut self, value: T) -> Result<(), T> {
        if self.len == self.buf.len() {
            return Err(value);
        }

        self.buf[self.len] = Some(value);
        self.len += 1;

        Ok(())
    }

    pub fn pop(&mut self) -> Option<T> {
        if self.len > 0 {
            let front = self.buf[0].take();
            for i in 1..self.len {
                self.buf.swap(i - 1, i);
            }
            self.len -= 1;
            front
        } else {
            None
        }
    }

    #[must_use]
    #[inline]
    pub const fn len(&self) -> usize {
        self.len
    }

    #[must_use]
    #[inline]
    pub const fn is_empty(&self) -> bool {
        self.len == 0
    }

    #[must_use]
    #[inline]
    pub const fn capacity(&self) -> usize {
        self.buf.len()
    }

    #[must_use]
    #[inline]
    pub const fn front(&self) -> Option<&T> {
        if self.len > 0 {
            self.buf[0].as_ref()
        } else {
            None
        }
    }

    #[must_use]
    #[inline]
    pub fn front_mut(&mut self) -> Option<&mut T> {
        if self.len > 0 {
            self.buf[0].as_mut()
        } else {
            None
        }
    }

    #[must_use]
    #[inline]
    pub const fn back(&self) -> Option<&T> {
        if self.len > 0 {
            self.buf[self.len - 1].as_ref()
        } else {
            None
        }
    }

    #[must_use]
    #[inline]
    pub fn back_mut(&mut self) -> Option<&mut T> {
        if self.len > 0 {
            self.buf[self.len - 1].as_mut()
        } else {
            None
        }
    }
}

impl<T: PartialEq> PartialEq for ArrayQueue<T> {
    fn eq(&self, other: &Self) -> bool {
        self.len == other.len && PartialEq::eq(&self.buf, &other.buf)
    }
}

impl<T: Eq> Eq for ArrayQueue<T> {}

impl<T: PartialOrd> PartialOrd for ArrayQueue<T> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        PartialOrd::partial_cmp(&self.buf, &other.buf)
    }
}

impl<T: Ord> Ord for ArrayQueue<T> {
    fn cmp(&self, other: &Self) -> Ordering {
        Ord::cmp(&self.buf, &other.buf)
    }
}

impl<T: Hash> Hash for ArrayQueue<T> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        Hash::hash(&self.buf, state);
    }
}

impl<T> FromIterator<T> for ArrayQueue<T> {
    fn from_iter<U: IntoIterator<Item=T>>(iter: U) -> Self {
        let vec: Vec<Option<T>> = iter.into_iter().map(|item| Some(item)).collect();
        Self {
            len: vec.len(),
            buf: vec.into_boxed_slice(),
        }
    }
}

impl<T: fmt::Debug> fmt::Debug for ArrayQueue<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(&self.buf, f)
    }
}

#[cfg(test)]
mod tests {
    use crate::array_queue::ArrayQueue;

    #[test]
    fn test_size() {
        assert_eq!(size_of::<ArrayQueue::<i32>>(), 24);
    }

    #[test]
    fn test_new() {
        let queue = ArrayQueue::<i32>::new(5);
        assert!(queue.is_empty());
    }

    #[test]
    fn test_push() {
        let mut queue = ArrayQueue::new(3);
        let ret = queue.push(0);
        assert!(ret.is_ok());
        assert_eq!(queue.front().copied(), Some(0));
        assert_eq!(queue.len(), 1);

        let ret = queue.push(1);
        assert!(ret.is_ok());
        assert_eq!(queue.front().copied(), Some(0));
        assert_eq!(queue.len(), 2);

        let ret = queue.push(2);
        assert!(ret.is_ok());
        let ret = queue.push(3);
        assert_eq!(ret, Err(3));
    }

    #[test]
    fn test_pop() {
        let mut queue = ArrayQueue::new(3);
        assert!(queue.push(0).is_ok());
        assert!(queue.push(1).is_ok());
        assert!(queue.push(2).is_ok());
        assert_eq!(queue.len(), 3);

        let ret = queue.pop();
        assert_eq!(ret, Some(0));
        assert_eq!(queue.len(), 2);
        assert_eq!(queue.front().copied(), Some(1));

        let ret = queue.push(3);
        assert!(ret.is_ok());
        assert_eq!(queue.len(), 3);

        assert_eq!(queue.pop(), Some(1));
        assert_eq!(queue.pop(), Some(2));
        assert_eq!(queue.pop(), Some(3));
        assert!(queue.is_empty());
    }
}